### PR TITLE
Endpoint for creating approximate GPS coordinates

### PR DESCRIPTION
**Pull request by Baalmart**
**_WHAT DOES THIS PR DO?_**

- [x] Creates a new endpoint for approximating GPS coordinates , provided the original GPS coordinates and the optional distance from original GPS point.
- [x] Calculate and add approximate coordinates to **Site** details during Site Creation process.
- [x] Utilise approximate coordinates when updating **Device** GPS coordinates during the Device deployment process.
- [x] Consequentially, update the **Sites** Schema to have the approximate coordinates accordingly

**_WHAT ISSUES ARE RELATED TO THIS PR?_**
- Jira cards
    [PLAT-1153]

**_HOW DO I TEST OUT THIS PR?_**
```
cd src/device-registry
```
`npm run dev-mac ` for LINUX
OR: 
`npm run dev-pc` for WINDOWS

**_WHICH ENDPOINTS SHOULD BE READY FOR TESTING?_** 

- [Approximate GPS coordinates](https://docs.airqo.net/airqo-platform-api/device-registry/create-sites#provide-approximate-gps-coordinates-based-on-distance-km-provided)
- [Create Site](https://app.gitbook.com/o/-MCogOVGQqjUharJKchL/s/-MKqyQkcfs54GYi-q96G/device-registry/create-sites#create-site)
- [Deploy Device](https://app.gitbook.com/o/-MCogOVGQqjUharJKchL/s/-MKqyQkcfs54GYi-q96G/device-registry/site-activities#deploy-device)

Screenshot 2022-06-10 at 20 08 36
Screenshot 2022-06-10 at 20 09 03
Screenshot 2022-06-10 at 20 11 19

[PLAT-1153]: https://airqoteam.atlassian.net/browse/PLAT-1153?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/device-registry/routes/api-v1.js [hunk 1]
```diff
@@ -2204,9 +2204,9 @@ router.post(
         })
         .bail()
         .customSanitizer((value) => {
-          return numeral(value).format("0.00000");
+          return numeral(value).format("0.00000000000000");
```
Comment (Baalmart) [inline src/device-registry/routes/api-v1.js:2207]: Updating the input validations for Site creation. Where Sanitization of the provided GPS coordinates is now changed from 5 decimal places to 14. "The more the decimal places, the better the accuracy of the coordinates".

File: src/device-registry/routes/api-v1.js [hunk 2]
```diff
@@ -2747,6 +2747,59 @@ router.get(
   siteController.findNearestSite
 );
 
+router.post(
+  "/sites/approximate",
+  oneOf([
+    [
```
Comment (Baalmart) [inline src/device-registry/routes/api-v1.js:2753]: The new route/endpoint definition for GPS coordinates approximation.

File: src/device-registry/utils/create-activity.js [hunk 3]
```diff
@@ -261,6 +262,16 @@ const createActivity = {
               3
             ),
           };
+
+          const responseFromCreateApproximateCoordinates = distance.createApproximateCoordinates(
+            { latitude, longitude }
+          );
```
Comment (Baalmart) [inline src/device-registry/utils/create-activity.js:268]: When creating an activity like deployment, also generate approximate coordinates using the provided ones and save them separately in the corresponding new fields in Site details..

File: src/device-registry/utils/create-site.js [hunk 4]
```diff
@@ -360,16 +360,43 @@ const manageSite = {
     try {
       const { body, query } = req;
       const { tenant } = query;
-      const { name, latitude, longitude, airqlouds } = body;
+      const {
+        name,
+        latitude,
```
Comment (Baalmart) [inline src/device-registry/utils/create-site.js:365]: The corresponding utils for site creation.

File: src/device-registry/utils/date.js [hunk 5]
```diff
@@ -126,6 +126,16 @@ const addDays = (number) => {
   }
 };
 
+const addMinutes = (number) => {
+  try {
```
Comment (Baalmart) [inline src/device-registry/utils/date.js:130]: New method for adding minutes to a date, although not used anywhere in the introduced functionality of this PR. Just here for future usage.

File: src/device-registry/utils/distance.js [hunk 6]
```diff
@@ -104,6 +99,89 @@ const distance = {
...
+        new Error("Number of decimal places must be at least 1.");
+      }
+      let value = Math.random() * (max - min + 1) + min;
+      return Number.parseFloat(value).toFixed(places);
+    }
+  },
+
+  createApproximateCoordinates: ({
+    latitude = 0,
+    longitude = 0,
```
Comment (Baalmart) [inline src/device-registry/utils/distance.js:146]: The util for creating approximate coordinates

File: src/device-registry/utils/generate-filter.js [hunk 7]
```diff
@@ -660,6 +662,14 @@ const generateFilter = {
       filter["_id"] = ObjectId(id);
     }
 
+    if (_id) {
```
Comment (Baalmart) [inline src/device-registry/utils/generate-filter.js:665]: adding more options (cases) when generating DB filters for Sites.

File: src/device-registry/models/Device.js [hunk 8]
```diff
@@ -29,6 +29,12 @@ const deviceSchema = new mongoose.Schema(
     longitude: {
       type: Number,
     },
+    approximate_distance_in_km: {
```
Comment (Baalmart) [inline src/device-registry/models/Device.js:32]: add the approximate distance from the actual point the device is situated.

File: src/device-registry/routes/api-v1.js [hunk 9]
```diff
@@ -2747,6 +2747,59 @@ router.get(
...
+          if (dp < 5) {
+            return Promise.reject(
+              "the latitude must have 5 or more characters"
+            );
+          }
+          return Promise.resolve("latitude validation test has passed");
+        })
+        .bail()
+        .customSanitizer((value) => {
+          return numeral(value).format("0.00000000000000");
```
Comment (Baalmart) [inline src/device-registry/routes/api-v1.js:2772]: Sanitising to 14 decimal places in the GPS coordinates input validation for the new approximation endpoint. "The more the decimal places, the better"

File: src/device-registry/routes/api-v1.js [hunk 10]
```diff
@@ -2747,6 +2747,59 @@ router.get(
...
+    [
+      body("latitude")
+        .exists()
+        .withMessage("the latitude should be provided")
+        .bail()
+        .matches(constants.LATITUDE_REGEX, "i")
+        .withMessage("the latitude provided is not valid")
+        .bail()
+        .custom((value) => {
+          let dp = decimalPlaces(value);
```
Comment (Baalmart) [inline src/device-registry/routes/api-v1.js:2762]: input validation to ensure that user provides at least 5 decimal places when inputing GPS values

**Review by NoahNsimbe — APPROVED** [review]
The changes LGTM. 
@Baalmart could you add tests for this functionality.

**Comment by Baalmart** [comment]
@sserurich , please take note of this PR, it is going into staging. There is an endpoint on it which you might want to take advantage of in the near future. :)

**Pull request merged**